### PR TITLE
User management

### PR DESCRIPTION
**Pull request by bogensberger**

### CODE REVIEW

File: CHANGES.txt [hunk 1]
```diff
@@ -21,6 +21,8 @@ Breaking Changes
 Changes
 =======
 
+ - Added support for `CREATE USER` and `DROP USER`
```
Comment (chaudum) [inline CHANGES.txt]: is this a sql feature? if so, please update the features file

File: blackbox/docs/sql/administration/user_management.txt [hunk 2]
```diff
@@ -0,0 +1,109 @@
+.. _administration_user_management:
```
Comment (chaudum) [inline blackbox/docs/sql/administration/user_management.txt]: if you want to link to the page instead of the headline, you can use the :doc: reference directive

File: sql/src/main/java/io/crate/action/sql/SQLOperations.java [hunk 3]
```diff
@@ -97,6 +97,13 @@ public Session createSession(@Nullable String defaultSchema, @Nullable String us
         return createSession(new SessionContext(defaultLimit, options, defaultSchema, userName));
     }
 
+    public Session createSession(@Nullable String defaultSchema, Set<Option> options, int defaultLimit, String username) {
```
Comment (chaudum) [inline sql/src/main/java/io/crate/action/sql/SQLOperations.java]: why did you create this new overloaded method? you could have just removed the @Nullable annotation from the userName argument of the method above

File: sql/src/main/java/io/crate/action/sql/SessionContext.java [hunk 4]
```diff
@@ -42,7 +42,19 @@ public SessionContext(Properties properties) {
...
+    public SessionContext(int defaultLimit,
+                          Set<Option> options,
+                          @Nullable String defaultSchema) {
+        this.defaultLimit = defaultLimit;
+        this.options = options;
+        this.userName = null;
+        setDefaultSchema(defaultSchema);
+    }
+
+    public SessionContext(int defaultLimit,
```
Comment (chaudum) [inline sql/src/main/java/io/crate/action/sql/SessionContext.java]: new constructor is not necessary imo

File: sql/src/main/java/io/crate/action/sql/SessionContext.java [hunk 5]
```diff
@@ -57,15 +69,15 @@ public String defaultSchema() {
         return defaultSchema;
     }
 
+    public void setDefaultSchema(@Nullable String schema) {
```
Comment (chaudum) [inline sql/src/main/java/io/crate/action/sql/SessionContext.java]: do not move existing code if not really necessary

File: users/src/main/java/io/crate/operation/user/UserManagerService.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+    static User CRATE_USER = new User("crate", true);
+
+    static {
+        MetaData.registerPrototype(TYPE, PROTO);
+    }
+
+    private final TransportCreateUserAction transportCreateUserAction;
+    private final TransportDropUserAction transportDropUserAction;
+    private final PermissionVisitor permissionVisitor = new PermissionVisitor();
```
Comment (chaudum) [inline users/src/main/java/io/crate/operation/user/UserManagerService.java]: can be static

File: users/src/main/java/io/crate/operation/user/UserManagerService.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    private class PermissionVisitor extends AnalyzedStatementVisitor<SessionContext, Boolean> {
+
+        private final SelectStatementPermissionVisitor selectVisitor = new SelectStatementPermissionVisitor();
+
+        @Override
+        protected Boolean visitAnalyzedStatement(AnalyzedStatement analyzedStatement,
+                                                 SessionContext sessionContext) {
+            return true;
```
Comment (chaudum) [inline users/src/main/java/io/crate/operation/user/UserManagerService.java]: I think @mfussenegger would prefer to deny all by default, so if you add new statements you are forced to implement the visit method for it. I see that point also, but don't have a strong opinion

File: users/src/main/java/io/crate/operation/user/UserManagerService.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    private class SelectStatementPermissionVisitor extends AnalyzedRelationVisitor<SessionContext, Boolean> {
+
+        @Override
+        protected Boolean visitAnalyzedRelation(AnalyzedRelation relation, SessionContext context) {
+            return true;
+        }
+
+        @Override
+        public Boolean visitQueriedTable(QueriedTable table, SessionContext context) {
+            return !(table.tableRelation().tableInfo().ident().name().equals("users") &&
```
Comment (chaudum) [inline users/src/main/java/io/crate/operation/user/UserManagerService.java]: hm, you have different behaviour here than in the PermissionVisitor. here you return true/false in the other case you throw

File: users/src/main/java/io/crate/operation/user/UserManagerService.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+                    sessionContext.userName(), analysis));
+            }
+            return true;
+        }
+
+        @Override
+        protected Boolean visitCreateUserStatement(CreateUserAnalyzedStatement analysis,
+                                                   SessionContext sessionContext) {
+            if (!isSuperUser(sessionContext.userName())){
+                throw new UnsupportedOperationException(String.format(Locale.ENGLISH, "User \"%s\" is not authorized to execute statement \"%s\"",
```
Comment (chaudum) [inline users/src/main/java/io/crate/operation/user/UserManagerService.java]: maybe new exception UnauthorizedOperationException?

File: users/src/main/java/io/crate/operation/user/UserManagerService.java (hunk 8 above)
Comment (bogensberger) [inline users/src/main/java/io/crate/operation/user/UserManagerService.java]: The PermissionVisitor has the AnalyzedStatement which is used in the exception message, that's why i let the throwing there. But i don't mind moving the throwing down and throw a different exception message.

File: users/src/main/java/io/crate/operation/auth/TrustAuthentication.java [hunk 10]
```diff
@@ -29,12 +30,16 @@
 public class TrustAuthentication implements AuthenticationMethod {
 
     static final String NAME = "trust";
-    private static final String SUPERUSER = "crate";
+    private final UserManager userManager;
+
+    public TrustAuthentication(UserManager userManager) {
```
Comment (mfussenegger) [inline users/src/main/java/io/crate/operation/auth/TrustAuthentication.java]: Could consider to reducing the dependency to a BooleanSupplier for the exists check. The AuthenticationMethod doesn't need to be able to access create/drop user functionality

File: users/src/main/java/io/crate/operation/auth/UserServiceFactoryImpl.java [hunk 11]
```diff
@@ -22,13 +22,41 @@
...
 
     @Override
-    public Authentication authService(Settings settings) {
-        return new HostBasedAuthentication(settings);
+    public Authentication authService(Settings settings, UserManager userManager) {
+        return new HostBasedAuthentication(settings, userManager);
+    }
+
+    @Override
+    public UserManager userManager(Settings settings,
```
Comment (mfussenegger) [inline users/src/main/java/io/crate/operation/auth/UserServiceFactoryImpl.java]: Maybe rename to `createUserManager` or `setupUserManager`- I'd also indicate with comments that this must only be called once as it registers transport actions

File: users/src/main/java/io/crate/operation/user/User.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    public User(String name, boolean superuser) {
+        this.name = name;
+        this.superuser = superuser;
+    }
+
+    public String name() {
+        return name;
+    }
+
+    public boolean superuser() {
```
Comment (mfussenegger) [inline users/src/main/java/io/crate/operation/user/User.java]: Usually we prefix boolean properties like this with `is`

File: users/src/main/java/io/crate/operation/user/User.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        return superuser;
+    }
+
+    public static Map<ColumnIdent, RowCollectExpressionFactory> sysUsersExpressions() {
+        return ImmutableMap.<ColumnIdent, RowCollectExpressionFactory>builder()
+            .put(SysUsersTableInfo.Columns.NAME, () -> new RowContextCollectorExpression<User, BytesRef>() {
+                @Override
+                public BytesRef value() {
+                    return new BytesRef(row.name());
```
Comment (mfussenegger) [inline users/src/main/java/io/crate/operation/user/User.java]: Could do the BytesRef creation in the User constructor

File: users/src/main/java/io/crate/operation/user/UserManagerService.java [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+    public UserManagerService(TransportCreateUserAction transportCreateUserAction,
+                              TransportDropUserAction transportDropUserAction,
+                              ClusterService clusterService) {
+        this.transportCreateUserAction = transportCreateUserAction;
+        this.transportDropUserAction = transportDropUserAction;
+        clusterService.add(this);
+    }
+
+    static Set<User> getUsersFromMetaData(@Nullable UsersMetaData metaData) {
```
Comment (mfussenegger) [inline users/src/main/java/io/crate/operation/user/UserManagerService.java]: I'd rename this to just `getUsers` - the `fromMetaData` is strange given that the argument can be null and you actually get a user back if it's null.

File: users/src/main/java/io/crate/operation/user/UserManagerService.java [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    @Override
+    public CompletableFuture<Long> dropUser(DropUserAnalyzedStatement analysis) {
+        FutureActionListener<WriteUserResponse, Long> listener = new FutureActionListener<>(WriteUserResponse::affectedRows);
+        transportDropUserAction.execute(new DropUserRequest(analysis.userName(), analysis.ifExists()), listener);
+        return listener;
+    }
+
+    public Iterable<User> userGetter() {
```
Comment (mfussenegger) [inline users/src/main/java/io/crate/operation/user/UserManagerService.java]: why `Getter`?

File: users/src/main/java/io/crate/operation/user/UserManagerService.java [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (!event.metaDataChanged()) {
+            return;
+        }
+        users = getUsersFromMetaData(event.state().metaData().custom(UsersMetaData.TYPE));
+    }
+
+
+    @Nullable
+    User findUser(@Nullable String userName) {
```
Comment (mfussenegger) [inline users/src/main/java/io/crate/operation/user/UserManagerService.java]: Does a call like `findUser(null)` really make sense?

File: users/src/main/java/io/crate/operation/user/UserManagerService.java (hunk 8 above)
Comment (mfussenegger) [inline users/src/main/java/io/crate/operation/user/UserManagerService.java]: Could just do equals on the ident using the SysUsersTableInfo.IDENT constant. 

Also - by reading the documentation it wasn't clear to me that the sys.users table can also only be read by a super user - the first paragraph is phrased in a way that sounds like only CREATE / DROP are special

Furthermore, this logic here won't work for any subqueries or joins. I'm also not sure if this is the right way to have permissions.
I'd have expected this to be the other way around . E.g. a table knows which permissions/roles are required - not that the UserManger knows about every table

File: users/src/main/java/io/crate/operation/user/UsersMetaData.java [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+
+    private UsersMetaData(List<String> users) {
+        this.users = users;
+    }
+
+    public static UsersMetaData newInstance(UsersMetaData instance) {
+        return new UsersMetaData(new ArrayList<>(instance.users));
+    }
+
+    static UsersMetaData of(String... user) {
```
Comment (mfussenegger) [inline users/src/main/java/io/crate/operation/user/UsersMetaData.java]: Can we remove this factory method and only have a constructor that takes the expected data structures without providing a convenience function that involves unnecessary transformations?

File: sql/src/main/java/io/crate/operation/user/UserManager.java [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+ */
+public interface UserManager {
+
+    /**
+     * creates a user
+     *
+     * @param analysis      analysed CREATE USER statement
+     * @return a future which returns the number of rows when the User is created
+     */
+    CompletableFuture<Long> createUser(CreateUserAnalyzedStatement analysis);
```
Comment (bogensberger) [inline sql/src/main/java/io/crate/operation/user/UserManager.java]: I agree on passing the user name directly.

The actual return value is a CompletableFuture and it fits quite well IMO. Especially in the dropUser case where the amount of rows varies between 0 and 1.

File: core/src/main/java/io/crate/module/CrateCoreModule.java [hunk 19]
```diff
@@ -53,6 +54,9 @@ public CrateCoreModule(Settings settings, IndexEventListenerProxy indexEventList
                 "but have not configured a license. Please request a license or deactivate the "+
                 "Enterprise Edition. https://crate.io/enterprise");
         }
+        if (CrateRestFilter.ES_API_ENABLED_SETTING.get(settings)) {
+            logger.warn("The Elasticsearch API is currently enabled, which makes it possible to request user data.");
```
Comment (bogensberger) [inline core/src/main/java/io/crate/module/CrateCoreModule.java]: It's not possible to read `sys.users` without superuser privileges. It's also not possible if you have no user at all as with HTTP.

File: blackbox/docs/sql/administration/user_management.txt [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+    | user_a      | FALSE     |
+    +-------------+-----------+
+    SELECT 3 rows in set (... sec)
+
+The column ``name`` shows the unique name of the user, the column ``superuser``
+shows whether the user has superuser privileges or not.
+
+.. warning::
+
+    When the :ref:`es_api_setting` is enabled, it is possible to read the
```
Comment (bogensberger) [inline blackbox/docs/sql/administration/user_management.txt]: It's not possible to query `sys.users` without superuser privileges. It's also not possible if you have no user at all as in HTTP. I think i need to point out that fact more clearly. The ES API is the only way to by-pass that and to read user data without any authentication and authorisation as superuser.

File: sql-parser/src/main/antlr/SqlBase.g4 [hunk 21]
```diff
@@ -68,6 +68,7 @@ statement
     | DROP SNAPSHOT qname                                                            #dropSnapshot
     | DROP FUNCTION (IF EXISTS)? name=qname
         '(' (functionArgument (',' functionArgument)*)? ')'                          #dropFunction
+    | DROP USER (IF EXISTS)? name=ident                                              #dropUser
```
Comment (bogensberger) [inline sql-parser/src/main/antlr/SqlBase.g4]: added it

File: sql/src/main/java/io/crate/action/sql/SQLOperations.java (hunk 3 above)
Comment (bogensberger) [inline sql/src/main/java/io/crate/action/sql/SQLOperations.java]: this method was is used in tests,  i remove it completly

File: users/src/main/java/io/crate/operation/auth/TrustAuthentication.java (hunk 10 above)
Comment (bogensberger) [inline users/src/main/java/io/crate/operation/auth/TrustAuthentication.java]: but the supplier can't take any arguments. In my case i need to pass the username.

File: sql/src/main/java/io/crate/operation/collect/sources/SysTableRegistry.java [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+    private final SystemCollectSource systemCollectSource;
+
+    @Inject
+    public SysTableRegistry(SysSchemaInfo sysSchemaInfo,
+                            SystemCollectSource systemCollectSource) {
+        this.sysSchemaInfo = sysSchemaInfo;
+        this.systemCollectSource = systemCollectSource;
+    }
+
+    public void registerSysTable(TableInfo tableInfo, Supplier<CompletableFuture<? extends Iterable<?>>> iterableSupplier,
```
Comment (bogensberger) [inline sql/src/main/java/io/crate/operation/collect/sources/SysTableRegistry.java]: I like that idea, i would like to this on a separate branch against master. Types weren't enforced so far and to do it here would make this already huge PR unnecessary huge.

File: users/src/main/java/io/crate/operation/auth/TrustAuthentication.java (hunk 10 above)
Comment (mfussenegger) [inline users/src/main/java/io/crate/operation/auth/TrustAuthentication.java]: But a `Predicate<String>` would work. The ctor would then just be called as `new TrustAuthentication(userManager::userExists)`

File: sql/src/main/java/io/crate/action/sql/SessionContext.java [hunk 23]
```diff
@@ -58,14 +60,21 @@ public String defaultSchema() {
...
     }
 
     public void setDefaultSchema(@Nullable String schema) {
         defaultSchema = MoreObjects.firstNonNull(schema, Schemas.DEFAULT_SCHEMA_NAME);
     }
 
+    /**
+     * set an authenticated user
+     */
+    public void setUser(User user) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/action/sql/SessionContext.java]: I liked having the User in the constructor better, why did you change  this to a setter? Afaik it doesn't make sense to change the user on a session?

File: sql/src/main/java/io/crate/analyze/AlterBlobTableAnalyzer.java [hunk 24]
```diff
@@ -41,7 +41,7 @@
     public AlterBlobTableAnalyzedStatement analyze(AlterBlobTable node, Row parameters) {
         TableIdent tableIdent = tableToIdent(node.table());
         assert BlobSchemaInfo.NAME.equals(tableIdent.schema()) : "schema name must be 'blob'";
-        BlobTableInfo tableInfo = (BlobTableInfo) schemas.getTableInfo(tableIdent);
+        BlobTableInfo tableInfo = (BlobTableInfo) schemas.getTableInfo(tableIdent, null);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/AlterBlobTableAnalyzer.java]: Why is it valid to pass null here?

File: sql/src/main/java/io/crate/analyze/DropBlobTableAnalyzer.java [hunk 25]
```diff
@@ -42,7 +42,7 @@ public DropBlobTableAnalyzedStatement analyze(DropBlobTable node) {
         BlobTableInfo tableInfo = null;
         boolean isNoop = false;
         try {
-            tableInfo = (BlobTableInfo) schemas.getTableInfo(tableIdent);
+            tableInfo = (BlobTableInfo) schemas.getTableInfo(tableIdent, null);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/DropBlobTableAnalyzer.java]: Same here - I think the user should be passed into getTableInfo instead of null

File: sql/src/main/java/io/crate/analyze/InsertFromValuesAnalyzer.java [hunk 26]
```diff
@@ -274,7 +280,7 @@ private void addValues(DocTableRelation tableRelation,
             final ColumnIdent columnIdent = column.ident().columnIdent();
             Object value;
             try {
-                valuesSymbol = valueNormalizer.normalizeInputForReference(valuesSymbol, column);
+                valuesSymbol = valueNormalizer.normalizeInputForReference(valuesSymbol, column, user);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/InsertFromValuesAnalyzer.java]: I think the valueNormalizer shouldn't require the user. All it does is retrieve References from the tableInfo - it currently re-retrieves the tableInfo but I think we could change that so that the tableInfo is passed into it as argument since the tableInfo is available here

File: sql/src/main/java/io/crate/operation/auth/UserServiceFactoryLoader.java [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+
+import io.crate.settings.SharedSettings;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.Iterator;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+
+public class UserServiceFactoryLoader {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/auth/UserServiceFactoryLoader.java]: please make the class final. It's still used as a base class but there is no need to inherit from it

File: sql/src/main/java/io/crate/operation/user/User.java [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.user;
+
+import java.util.Set;
+
+public interface User {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/user/User.java]: Would it make sense to change this to a class? `new User(name, roles)`

File: sql/src/main/java/io/crate/analyze/AlterBlobTableAnalyzer.java [hunk 29]
```diff
@@ -38,11 +39,10 @@
         this.schemas = schemas;
     }
 
-    public AlterBlobTableAnalyzedStatement analyze(AlterBlobTable node, Row parameters) {
+    public AlterBlobTableAnalyzedStatement analyze(AlterBlobTable node, Row parameters, SessionContext sessionContext) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/AlterBlobTableAnalyzer.java]: Please only require the user if nothing else is required from the sessionContext

File: sql/src/main/java/io/crate/analyze/InsertFromValuesAnalyzer.java [hunk 30]
```diff
@@ -253,7 +261,8 @@ private void addValues(DocTableRelation tableRelation,
                            ReferenceToLiteralConverter refToLiteral,
                            int numPrimaryKeys,
                            Function<List<BytesRef>, String> idFunction,
-                           int bulkIdx) throws IOException {
+                           int bulkIdx,
+                           User user) throws IOException {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/InsertFromValuesAnalyzer.java]: User param is now unused?

File: sql/src/main/java/io/crate/analyze/UpdateAnalyzer.java [hunk 31]
```diff
@@ -168,7 +160,8 @@ private void analyzeAssignment(Assignment node,
                                    ExpressionAnalyzer expressionAnalyzer,
                                    ExpressionAnalyzer columnExpressionAnalyzer,
                                    ExpressionAnalysisContext expressionAnalysisContext,
-                                   TransactionContext transactionContext) {
+                                   TransactionContext transactionContext,
+                                   User user) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/UpdateAnalyzer.java]: unused User param

File: sql/src/main/java/io/crate/metadata/doc/DocTableInfo.java [hunk 32]
```diff
@@ -371,6 +373,11 @@ public TableParameterInfo tableParameterInfo() {
         return supportedOperations;
     }
 
+    @Override
+    public Set<User.Role> requiredUserRoles() {
+        return ImmutableSet.of();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/doc/DocTableInfo.java]: Could add this as default implementation on the interface itself

File: sql/src/main/java/io/crate/operation/user/UserManager.java [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+     * deletes a user
+     * @param analysis      analysed DROP USER statement
+     * @return a future which returns the number of rows when the User is dropped
+     */
+    CompletableFuture<Long> dropUser(DropUserAnalyzedStatement analysis);
+
+    /**
+     * checks if user is allowed to execute statement
+     * @param analysis          analysed statement
+     * @param sessionContext    current session context
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/user/UserManager.java]: maybe document the exceptions using `@throws`, could also be renamed to something like `raiseIfUnauthorized` / `ensureAuthorized` - to be more in line with our usual naming for methods that validate & raise if not valid

File: sql/src/main/java/io/crate/operation/user/UserManagerProvider.java [hunk 34]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    }
+
+
+    @Override
+    public UserManager get() {
+        return userManager;
+    }
+
+    private static class NoopUserManager implements UserManager {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/user/UserManagerProvider.java]: maybe rename to `UnsupportedUserManager` or something like that

File: sql/src/test/java/io/crate/metadata/SchemasTest.java [hunk 35]
```diff
@@ -73,6 +92,48 @@ public void setUp() throws Exception {
...
+            public Routing getRouting(WhereClause whereClause, @Nullable String preference) {
+                return null;
+            }
+
+            @Override
+            public Set<User.Role> requiredUserRoles() {
+                return ImmutableSet.of(User.Role.SUPERUSER);
+            }
+        });
+        sysSchemaInfo.registerSysTable(new StaticTableInfo(unauthorized,
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/metadata/SchemasTest.java]: Would it be possible to use one of the existing sys tables here instead?

File: users/src/main/java/io/crate/operation/user/UserManagerService.java (hunk 7 above)
Comment (mfussenegger) [inline users/src/main/java/io/crate/operation/user/UserManagerService.java]: Usually yes, in this case it's okay for me because we've another security check due to the requiredRoles on the tables

File: CHANGES.txt (hunk 1 above)
Comment (bogensberger) [inline CHANGES.txt]: no
